examples/arm/executor_runner/arm_perf_monitor.cpp

```diff
@@ -0,0 +1,173 @@
+ */
+
+#include <cinttypes>
+#include <vector>
+
+#include "arm_perf_monitor.h"
+
+#ifdef ETHOSU
+#include <ethosu_driver.h>
+#include <executorch/runtime/platform/log.h>
+#include <pmu_ethosu.h>
+
+static uint32_t ethosu_inference_count = 0;
+static uint64_t ethosu_ArmBackendExecuteCycleCountStart = 0;
+static uint64_t ethosu_ArmBackendExecuteCycleCount = 0;
+static uint64_t ethosu_ArmWhenNPURunCycleCountStart = 0;
+static uint64_t ethosu_ArmWhenNPURunCycleCount = 0;
+static uint64_t ethosu_pmuCycleCount = 0;
+static std::vector<uint64_t> ethosu_pmuEventCounts(
+    ETHOSU_PMU_Get_NumEventCounters(),
+    0);
+
+static const uint32_t ethosu_pmuCountersUsed = 4;
+// ethosu_pmuCountersUsed should match numbers of counters setup in
+// ethosu_inference_begin() and not be more then the HW supports
+static_assert(ETHOSU_PMU_NCOUNTERS >= ethosu_pmuCountersUsed);
+
+extern "C" {
+
+// Callback invoked at start of NPU execution
+void ethosu_inference_begin(struct ethosu_driver* drv, void*) {
+  // Enable PMU
+  ETHOSU_PMU_Enable(drv);
+  ETHOSU_PMU_PMCCNTR_CFG_Set_Stop_Event(drv, ETHOSU_PMU_NPU_IDLE);
+  ETHOSU_PMU_PMCCNTR_CFG_Set_Start_Event(drv, ETHOSU_PMU_NPU_ACTIVE);
+
+  // Setup 4 counters
+  ETHOSU_PMU_Set_EVTYPER(drv, 0, ETHOSU_PMU_AXI0_RD_DATA_BEAT_RECEIVED);
+  ETHOSU_PMU_Set_EVTYPER(drv, 1, ETHOSU_PMU_AXI1_RD_DATA_BEAT_RECEIVED);
+  ETHOSU_PMU_Set_EVTYPER(drv, 2, ETHOSU_PMU_AXI0_WR_DATA_BEAT_WRITTEN);
+  ETHOSU_PMU_Set_EVTYPER(drv, 3, ETHOSU_PMU_NPU_IDLE);
+  // Enable 4 counters
+  ETHOSU_PMU_CNTR_Enable(drv, 0xf);
+
+  ETHOSU_PMU_CNTR_Enable(drv, ETHOSU_PMU_CCNT_Msk);
+  ETHOSU_PMU_CYCCNT_Reset(drv);
+
+  // Reset all counters
+  ETHOSU_PMU_EVCNTR_ALL_Reset(drv);
+
+  // Save Cortex-M cycle clock to calculate total CPU cycles used in
+  // ethosu_inference_end()
+  ethosu_ArmWhenNPURunCycleCountStart = ARM_PMU_Get_CCNTR();
+}
+
+// Callback invoked at end of NPU execution
+void ethosu_inference_end(struct ethosu_driver* drv, void*) {
+  ethosu_inference_count++;
+  ethosu_pmuCycleCount += ETHOSU_PMU_Get_CCNTR(drv);
+
+  for (size_t i = 0; i < ethosu_pmuCountersUsed; i++) {
+    ethosu_pmuEventCounts[i] += ETHOSU_PMU_Get_EVCNTR(drv, i);
+  }
+  ETHOSU_PMU_Disable(drv);
+  // Add Cortex-M cycle clock used during this NPU execution
+  ethosu_ArmWhenNPURunCycleCount +=
+      (ARM_PMU_Get_CCNTR() - ethosu_ArmWhenNPURunCycleCountStart);
+}
+
+// Callback invoked at start of ArmBackend::execute()
+void ArmBackend_execute_begin() {
+  // Save Cortex-M cycle clock to calculate total CPU cycles used in
+  // ArmBackend_execute_end()
+  ethosu_ArmBackendExecuteCycleCountStart = ARM_PMU_Get_CCNTR();
+}
+
+// Callback invoked at end of ArmBackend::execute()
+void ArmBackend_execute_end() {
+  // Add Cortex-M cycle clock used during this ArmBackend::execute()
+  ethosu_ArmBackendExecuteCycleCount +=
+      (ARM_PMU_Get_CCNTR() - ethosu_ArmBackendExecuteCycleCountStart);
+}
+}
+
+void StartMeasurements() {
```

Comment:
can we add some more logic here to differentiate between two pairs of stat-stop? i.e.
```
start1
op1
stop1
...
start2
op2
stop2
...
```

Also passing some `id` like information in the pmu dump for post process.

Reply:
I tried to keep this simple and hope to get the ET devtools to do some of this things, but I like the idea of some id at least then logic could be added later. Right now they are not supposed to be called more then around each time the NPU is running, so never more then one at the time, but it might be nice anyway.